lib/file-source/src/checkpointer.rs

```diff
@@ -103,7 +177,67 @@ impl Checkpointer {
         Ok(self.checkpoints.len())
     }
 
+    /// Read persisted checkpoints from disk, preferring the new JSON file format but falling back
+    /// to the legacy system when those files are found instead.
     pub fn read_checkpoints(&mut self, ignore_before: Option<time::SystemTime>) {
+        // First try reading from the tmp file location. If this works, it means that the previous
+        // process was interrupted in the process of checkpointing and the tmp file should contain
+        // more recent data that should be preferred.
+        match self.read_checkpoints_file(&self.tmp_file_path) {
+            Ok(state) => {
+                warn!(message = "Recovered checkpoint data from interrupted process.");
+                self.set_state(state);
+
+                // Try to move this tmp file to the stable location so we don't immediately overwrite
+                // it when we next persist checkpoints.
+                if let Err(error) = fs::rename(&self.tmp_file_path, &self.stable_file_path) {
+                    warn!(message = "Error persisting recovered checkpoint file.", %error);
+                }
+                return;
+            }
+            Err(error) if error.kind() == io::ErrorKind::NotFound => {
+                // This is expected, so no warning needed
+            }
+            Err(error) => {
+                error!(message = "Unable to recover checkpoint data from interrupted process.", %error);
+            }
+        }
+
+        // Next, attempt to read checkpoints from the stable file location. This is the
+        // expected location, so warn more aggressively if something goes wrong.
+        match self.read_checkpoints_file(&self.stable_file_path) {
+            Ok(state) => {
+                info!(message = "Loaded checkpoint data.");
+                self.set_state(state);
+                return;
+            }
+            Err(error) if error.kind() == io::ErrorKind::NotFound => {
+                // This is expected, so no warning needed
+            }
+            Err(error) => {
+                warn!(message = "Unable to load checkpoint data.", %error);
+                return;
+            }
+        }
+
+        // If we haven't returned yet, go ahead and look for the legacy files and try to read them.
+        info!("Attempting to read legacy checkpoint files.");
+        self.read_legacy_checkpoints(ignore_before);
+
+        if let Ok(_) = self.write_checkpoints() {
+            fs::remove_dir_all(&self.directory).ok();
+        }
+    }
+
+    // TODO: Do we have the equivalent of ignore_before here? How should we go about expiring
```

Comment:
Is this left for a future PR?

Reply:
Initially it was, but now that I'm thinking about it more I think this needs to be handled so that the file doesn't grow forever.

Reply:
Perhaps we can chunk the checkpoint files? Expire by removing old ones when a new chunk is finalized?